File: Java/QCAndroid/src/org/quickconnectfamily/QuickConnect.java
/*
 Copyright (c) 2008, 2009, 2011 Lee Barney
 Permission is hereby granted, free of charge, to any person obtaining a 
 copy of this software and associated documentation files (the "Software"), 
 to deal in the Software without restriction, including without limitation the 
 rights to use, copy, modify, merge, publish, distribute, sublicense, 
 and/or sell copies of the Software, and to permit persons to whom the Software 
 is furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be 
 included in all copies or substantial portions of the Software.
 
 
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
 INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
 PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
 CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
 OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 
 */
package org.quickconnectfamily;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.concurrent.RejectedExecutionHandler;
import java.util.concurrent.SynchronousQueue;
import java.util.concurrent.ThreadPoolExecutor;
import java.util.concurrent.TimeUnit;
import java.util.concurrent.atomic.AtomicInteger;

import android.os.Handler;

/**
 * The QuickConnect class is used to create control stacks and execute them.
 * Control stacks consist a few types of objects that all implement the
 * ControlObject interface. These types are:
 * <ol>
 * <li><b>ValCO</b> Validation Control Objects - these stacks objects are used
 * to validate values that will be used later in other control objects. The
 * items validated are usually passed in as parameters. By convention names for
 * objects of this type end with ValCO.</li>
 * <li><b>DCO</b> Data Control Objects - these stack objects are used to
 * interact with databases, remote servers, files, and are also used to do data
 * manipulation/computation. By convention names for objects of this type end
 * with DCO.</li>
 * <li><b>VCO</b>View Control Objects - these stack objects are used to do user
 * interface updates.</li>
 * <li><b>ECO</b>Error Control Objects - these stack objects are used to notify
 * the user of errors and do any view cleanup required if an bad data or an
 * error/exception occurs during the execution of a stack. By convention names
 * for objects of this type end with VCO.</li>
 * </ol>
 * 
 * An example of a login stack could include the following control objects:
 * <ul>
 * <li><b>CredentialsValCO</b> - the handleIt method performs length/content
 * validation on the user name and password. Returns either QC.STACK_EXIT if 
 * validation fails or QC.STACK_CONTINUE on passing validation</li>
 * <li><b>GetUserCO</b> - the handleIt method queries the database to see if a
 * user exists with the validated user name and password. Adds the user to the parameters map 
 * and returns the QC.STACK_CONTINUE or calls handleError and returns QC.STACK_EXIT.</li>
 * <li><b>LoginSuccessVCO</b> - the handleIt method updates the UI.</li>
 * 
 * @author Lee S. Barney
 * 
 */
public class QuickConnect {

	private static HashMap<String, ArrayList<ControlObject>> validationMap;
	private static HashMap<String, ArrayList<ControlObject>> dataMap;
	private static HashMap<String, ArrayList<ControlObject>> viewMap;
	private static HashMap<String, ArrayList<ControlObject>> errorMap;
	private static HashMap<String, ArrayList<String>> groupMap;


	private static Handler theMainThreadHandler;
	private static ThreadPoolExecutor thePool;
	private static Exception error;

	static {
		validationMap = new HashMap<String, ArrayList<ControlObject>>();
		dataMap = new HashMap<String, ArrayList<ControlObject>>();
		viewMap = new HashMap<String, ArrayList<ControlObject>>();
		errorMap = new HashMap<String, ArrayList<ControlObject>>();
		groupMap = new HashMap<String, ArrayList<String>>();
		try {
			SynchronousQueue<Runnable> worksQueue = new SynchronousQueue<Runnable>(
					true);
			RejectedExecutionHandler executionHandler = new PoolRejectedExecutionHandler();

			thePool = new ThreadPoolExecutor(3, Integer.MAX_VALUE, 10, TimeUnit.SECONDS,
					worksQueue, executionHandler);
			thePool.allowCoreThreadTimeOut(true);
			theMainThreadHandler = new Handler();
		} catch (Exception e) {
			thePool = null;
			error = e;
		}
	}

	/**
	 * This method is the Control Object stack trigger. Call it from anywhere in your
	 * user interface code to execute a stack created using the mapCommandTo** methods.
	 * 
	 * @param command
	 *            - the command string that is the key associated with the
	 *            desired Command Object stack.
	 * @param parameters
	 *            - a HashMap containing any key-value pairs you wish to pass to
	 *            all Command Objects' in the stack. If User Interface elements 
	 *            are to be added as values, a WeakReference to that element 
	 *            must be passed instead. This mutable HashMap 
	 *            appears as the 'parameters' value passed into all Command
	 *            Objects in the stack.
	 */
	
	public static void handleRequest(String command, HashMap<Object, Object>parameters){
		QuickConnect.handleRequest(command, parameters, true, null);
		
	}
	
	public static void handleRequest(String command, HashMap<Object, Object> parameters, boolean runInBackground, StackCallback aCallback) {
		if(thePool == null){
			error.printStackTrace();
		}
		ArrayList<String> subCommands = groupMap.get(command);
		if(subCommands != null){
			int numSubCommands = subCommands.size();
			for(int i = 0; i < numSubCommands; i++){
				QuickConnect.handleRequest(subCommands.get(i), parameters, runInBackground, aCallback);
			}
			return;
		}
		AtomicInteger numRequestsToTrack = new AtomicInteger(1);
		QuickConnect.handleRequest(command, parameters, runInBackground, aCallback, numRequestsToTrack);
		
	}
	
	public static void handleRequest(String[] commands, HashMap<Object, Object>parameters, boolean runInBackground, StackCallback aCallback){
		int numCommands = commands.length;
		AtomicInteger numRequestsToTrack = null;
		if(runInBackground){
			numRequestsToTrack = new AtomicInteger(1);
		}
		else{
			numRequestsToTrack = new AtomicInteger(numCommands);
		}
		for(int i = 0; i < numCommands; i++){
			QuickConnect.handleRequest(commands[i], parameters, runInBackground, aCallback, numRequestsToTrack);
		}
	}
	
	public static void handleRequest(CommandDescriptor aDescriptor){
		QuickConnect.handleRequest(aDescriptor.getCommand(), aDescriptor.getData(), aDescriptor.runInBackground(), aDescriptor.getCallback());
	}
	
	public static void handleBatchRequest(CommandDescriptor[] theDescriptors){
		int numDescriptors = theDescriptors.length;
		AtomicInteger numRequestsToTrack = new AtomicInteger(numDescriptors);
		for(int i = 0; i < numDescriptors; i++){
			CommandDescriptor aDescriptor = theDescriptors[i];
			QuickConnect.handleRequest(aDescriptor.getCommand(), aDescriptor.getData(), aDescriptor.runInBackground(), aDescriptor.getCallback(), numRequestsToTrack);
		}
	}

	private static void handleRequest(String command, HashMap<Object, Object> parameters, boolean runInBackground, StackCallback aCallback, AtomicInteger numRequestsToTrack) {
		if(runInBackground){
			thePool.execute(new ControlObjectStack(command, parameters, aCallback, numRequestsToTrack, theMainThreadHandler));
		}
		else{
			new ControlObjectStack(command, parameters, aCallback, numRequestsToTrack, theMainThreadHandler).run();
		}
	}

	/**
	 * This method is similar to handleRequest but is used to trigger stacks
	 * made of Error Control Objects. Call it from anywhere in your Control Object code where
	 * you wish to trigger a specified error handling stack. It is common to
	 * call this method from within Validation Control Objects if validation
	 * fails prior to returning QC.STACK_EXIT to terminate execution of the stack. It is
	 * also common to call this method from within Data Control Objects when
	 * data retrieved from remote sources or local databases is not what it
	 * should be.
	 * 
	 * @param command
	 *            - the command string that is the key associated with the
	 *            desired Error Command Object stack.
	 * @param parameters
	 *            - an ArrayList containing any values you wish to pass to all
	 *            Error Command Objects' in the stack. This ArrayList appears as
	 *            the 'paramters' value passed into all Error Command Objects in
	 *            the stack.
	 */
	public static void handleError(String command,HashMap<Object, Object> parameters) {
		new ControlObjectStack(command, parameters, null, new AtomicInteger(1), theMainThreadHandler).dispatchToECO(command,parameters);
	}

	protected static HashMap<String, ArrayList<ControlObject>> getValidationMap() {
		return validationMap;
	}

	protected static HashMap<String, ArrayList<ControlObject>> getDataMap() {
		return dataMap;
	}

	protected static HashMap<String, ArrayList<ControlObject>> getViewMap() {
		return viewMap;
	}

	protected static HashMap<String, ArrayList<ControlObject>> getErrorMap() {
		return errorMap;
	}

	/**
	 * This method maps a specific command to a Data Control Object (DCO).
	 * DCOs are used to retrieve data from remote sources, local databases, and
	 * do data manipulation/calculation.
	 * 
	 * @param command
	 *            - the key used to identify the stack to which the specific DCO
	 *            is to be mapped.
	 * @param controlObjectClass
	 *            - the class of DCO to be added to the stack. There may be as
	 *            many different kinds of these added as needed.
	 */
	public static void mapCommandToDCO(String command, Class<?> controlObjectClass) {
		mapCommandToControlObject(command, controlObjectClass, dataMap);
	}

	/**
	 * This method maps a specific command to a View Control Object (VCO). VCOs
	 * are used to update the user interface based on the data collected and
	 * calculated in the DCOs in the same stack. VCOs are executed in the UI
	 * thread for Android and Java SE implementations. No threading is done for
	 * Java EE since Java EE applications are already threaded.
	 * 
	 * @param command
	 *            - the key used to identify the stack to which the specific VCO
	 *            is to be mapped.
	 * @param controlObjectClass
	 *            - the class of VCO to be added to the stack. There may be as
	 *            many different kinds of these added as needed.
	 */
	public static void mapCommandToVCO(String command, Class<?> controlObjectClass) {
		mapCommandToControlObject(command, controlObjectClass, viewMap);
	}

	/**
	 * This method maps a specific command to a Validation Control Object
	 * (ValCO). ValCOs are used to validate data before the DCOs and VCOs are
	 * called.
	 * 
	 * @param command
	 *            - the key used to identify the stack to which the specific
	 *            ValCO is to be mapped.
	 * @param controlObjectClass
	 *            - the class of ValCO to be added to the stack. There may be as
	 *            many different kinds of these added as needed.
	 */
	public static void mapCommandToValCO(String command, Class<?> controlObjectClass) {
		mapCommandToControlObject(command, controlObjectClass, validationMap);
	}

	/**
	 * This method maps a specific command to a Error Control Object (ECO). ECOs
	 * are used to notify the user of errors and do any cleanup necessary. ECOs
	 * are executed in the UI thread for Android and Java SE implementations. No
	 * threading is done for Java EE since Java EE applications are already
	 * threaded.
	 * 
	 * @param command
	 *            - the key used to identify the stack to which the specific ECO
	 *            is to be mapped.
	 * @param controlObjectClass
	 *            - the class of ECO to be added to the stack. There may be as
	 *            many different kinds of these added as needed.
	 */
	public static void mapCommandToECO(String command, Class<?> controlObjectClass) {
		mapCommandToControlObject(command, controlObjectClass, errorMap);
	}

	private static void mapCommandToControlObject(String command, Class<?> controlObjectClass,
			HashMap<String, ArrayList<ControlObject>> aMap) {
		if (command != null) {
			ArrayList<ControlObject> handlers = null;
			if ((handlers = aMap.get(command)) == null) {
				handlers = new ArrayList<ControlObject>();
				aMap.put(command, handlers);
			}
			try {
				handlers.add((ControlObject) (controlObjectClass.newInstance()));
			} catch (Exception e) {
				String message = "Unable to create object of type: "
						+ controlObjectClass.getCanonicalName() + ".  "
						+ e.getCause();
				if (e.getMessage() != null) {
					message = e.getMessage();
				}
				System.out.println(message);
			}
		}

	}
	
	public void mapCommandToSubCommands(String command, String[] subCommands) throws Exception{
		if(command == null || subCommands == null){
			throw new Exception("Error: Missing a command. Both the command and subCommands must not be null.");
		}
		int numCommands = subCommands.length;
		for(int i = 0; i < numCommands; i++){
			addSubCommand(command, subCommands[i]);
		}
	}
	
	private void addSubCommand(String aCommand, String aSubCommand) throws Exception{
		if(aCommand == null || aSubCommand == null){
			throw new Exception("Error: attempting to add a null sub-command to "+aCommand);
		}
		ArrayList<String> subCommands = groupMap.get(aCommand);
		if(subCommands == null){
			subCommands = new ArrayList<String>();
			groupMap.put(aCommand, subCommands);
		}
		subCommands.add(aSubCommand);
	}
}

File: JavaScript/tests/web_worker_test/QuickConnect.js
/*
 Copyright (c) 2008, 2009, 2011, 2012 Lee Barney
 Permission is hereby granted, free of charge, to any person obtaining a 
 copy of this software and associated documentation files (the "Software"), 
 to deal in the Software without restriction, including without limitation the 
 rights to use, copy, modify, merge, publish, distribute, sublicense, 
 and/or sell copies of the Software, and to permit persons to whom the Software 
 is furnished to do so, subject to the following conditions:
 
 The above copyright notice and this permission notice shall be 
 included in all copies or substantial portions of the Software.
 
 
 
 
 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR IMPLIED, 
 INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY, FITNESS FOR A 
 PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT 
 HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION OF 
 CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE 
 OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
 
 
 */

var qc = new Object();

qc.WAIT_FOR_DATA = 'wAiT';
qc.STACK_EXIT = 'ExIt_StAcK';
qc.STACK_CONTINUE = 'ContInUe';
 
var validationMap = new Object()
var dataMap = new Object()
var viewMap = new Object()
var errorMap = new Object()
var groupMap = new Object()

qc.validationMapConsumables = new Object()
qc.dataMapConsumables = new Object()
qc.viewMapConsumables = new Object()


qc.missingCommandError = 0
qc.success = 1

qc.workerThreadMap = new Object()
qc.customFunctionMap = new Object()

function debug(aMessage){
  console.log(aMessage)
}

 //generate a UUID
qc.generateUUID = function(){
    var uuid = 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
                                                              var r = Math.random()*16|0, v = c == 'x' ? r : (r&0x3|0x8);
                                                              return v.toString(16);
                                                              })
    return uuid
}

qc.RequestHandlerPool = function(min_size) {
    var workerQueue = [];
    var workingThreads = {}
    for (var i = 0 ; i < min_size ; i++) {
        workerQueue.push(new qc.HandlerThread(this))
    }
    this.executeHandleRequestTask = function(handleRequestData, continuationFunction) {
        var workerThread = null;
        if (workerQueue.length > 0) {
            // get the worker from the front of the queue
            workerThread = workerQueue.shift()
            //console.log('getting worker: '+workerThread)
        } else {
            // no free workers, create a new one
            workerThread = new qc.HandlerThread()
            //console.log('created worker: '+workerThread)
        }
        workingThreads[workerThread.uuid] = workerThread
        workerThread.run(handleRequestData, continuationFunction);
    }
    this.reuseWorkerThread = function(aWorkerThread){
      delete workingThreads[aWorkerThread.uuid]
      workerQueue.push(aWorkerThread)
    }
    this.freeUnusedHandlerThreads = function(){
      var currentTime = new Date().getTime()
      while(workerQueue.length > min_size && currentTime - workerQueue[0].usedTime > 30000){
        workerQueue.shift()
      }
    }
}

function clearOldHandlerThreads(){
  requestPool.freeUnusedHandlerThreads();
}

setInterval(clearOldHandlerThreads, 60000);

qc.HandlerThread = function() {
    var self = this
    this.theWorker = null
    this.requestData = null
    this.run = function(handleRequestData, continuationFunction) {
        if(!this.theWorker){
          this.workerFileName = 'RequestWorker.js'
          if(handleRequestData.data.workerFileName){
            this.workerFileName = handleRequestData.data.workerFileName
          }
          this.theWorker = new Worker(this.workerFileName)
        }
        this.requestData = handleRequestData
        var uuid = qc.generateUUID()
        handleRequestData.data.stackID = uuid
        qc.workerThreadMap[uuid] = this
        // create a new web worker
        this.theWorker.continuationFunction = continuationFunction;
        this.theWorker.onmessage = function(event) {
            //call back from worker
            var results = event.data
            if(results.constructor == String){
              results = JSON.parse(results)
            }
            if(results.stackFlag == qc.STACK_CONTINUE){
                //free the thread early so the it is available before view code runs
                requestPool.reuseWorkerThread(self)
                var result = qc.dispatchToVCF(results.cmd, results.data)
                if(this.continuationFunction){
                  this.continuationFunction(results.data)
                }
                else if(this.continuationFunction){
                  window[this.continuationFunction](results.data)
                }
            }
            else if(results.stackFlag == qc.STACK_EXIT){
                requestPool.reuseWorkerThread(self)
            }
            else if(results.log){
              console.log(results.log)
            }
            else if(results.custom){
              var customCallData = results.custom
              qc.customFunctionMap[customCallData.cmd](customCallData.params)
            }
            else if(results.startStack){
              var allStacksCompleteCallbackFunc = eval(results.allStacksCompleteCallback)
              qc.handleRequest(results.aCommandArray, results.requestData, allStacksCompleteCallbackFunc, results.runParallel)
            }
            else if(results.err){
              console.error(results.err)
            }
            else if(results.warn){
              console.warn(results.warn)
            }
            else if(results.alert){
              alert(results.alert)
            }
            else{
              console.log('Error: unknown push to main page thread.')
            }
        }
        //console.log('posting to worker: ')
        this.theWorker.postMessage(JSON.stringify(handleRequestData))
    }
}


qc.continueStack = function(stackID, responseID, responseData){
  var theThread = qc.workerThreadMap[stackID]
  theThread.requestData.data[responseID] = responseData
  theThread.requestData.data['continue'] = true
  theThread.theWorker.postMessage(JSON.stringify(theThread.requestData))
}

var requestPool = new qc.RequestHandlerPool(1)

qc.handleRequest = function(aCommandArray, requestData
                    , allStacksCompleteCallback, runParallel) {
  if (!aCommandArray || aCommandArray.length == 0){
    console.warn('WARNING: attempting to execute a request without one or more commands.')
  }
  if (Array.isArray(requestData)){
    requestData = {'paramArray':requestData}
  }
  if (!Array.isArray(aCommandArray)) {
    aCommandArray = [aCommandArray]
  }
  requestData = requestData || new Object()
  var calledFromWithinStack = false;
  if(requestData.stackID){
    delete requestData['stackID']
    calledFromWithinStack = true;
  }/*
  if(!calledFromWithinStack){
    var aFunc = arguments.callee.caller
    while(aFunc){
      if(aFunc.qc_stack_command){
        calledFromWithinStack = true
        break
      }
      aFunc = arguments.callee.caller
    }
  }*/
  if(calledFromWithinStack){
    delete requestData['thisRequestWorker']
    var postData = {'startStack':true, 'aCommandArray':aCommandArray,
              'requestData':requestData,
              'runParallel':runParallel}
    theWorker.postMessage(JSON.stringify(postData))
    return;
  }
  if (runParallel){
    var numCommands = aCommandArray.length
    for (var i = 0; i < numCommands; i++){
      var aCmd = aCommandArray.shift()
      var subCommands = groupMap[aCmd]
      if(subCommands != null){
          var numSubCommands = subCommands.length
          qc.handleRequest(subCommands.slice(), requestData, allStacksCompleteCallback, runParallel)
          return;
      }
      requestPool.executeHandleRequestTask({'cmd':aCmd, 'data':requestData}, allStacksCompleteCallback)
    }
  }
  else{
    function executeNextCommand(){
      var aCommand = aCommandArray.shift()
      requestPool.executeHandleRequestTask({'cmd':aCommand, 'data':requestData},
          aCommandArray.length > 1 ? executeNextCommand : allStacksCompleteCallback)
    }
      //shift the first command off of the array so the stack 
      //will not be called for it twice.
      var aCommand = aCommandArray.shift()
      var subCommands = groupMap[aCommand]
      if(subCommands != null){
          var numSubCommands = subCommands.length
          for(var i = 0; i < numSubCommands; i++){
              qc.handleRequest(subCommands, requestData, allStacksCompleteCallback, runParallel)
          }
          return;
      }
      requestPool.executeHandleRequestTask({'cmd':aCommand, 'data':requestData},
              aCommandArray.length > 0 ? executeNextCommand : allStacksCompleteCallback)
  }
}

qc.handleBatchRequest = function(requestDefinitions){
  var numPairs = arguments.length
  for(var i = 0; i < numPairs; i++){
    var aDefinition = arguments[i]
    //all commands are run parallel in a batch
    qc.handleRequest(aDefinition.cmd, aDefinition.parameters, aDefinition.callback, true)
  }
}

qc.cloneConsumableStacks = function(aCmd, uuid){
  //console.log('------------------in cloning: '+aCmd+"---------------------------")
  //if mappings are found then duplicate the mapped 
  //control function arrays for consumption

  if (!validationMap[aCmd] && !dataMap[aCmd] 
          && !viewMap[aCmd] && !errorMap[aCmd]) {
    //console.log('no stacks found')
    return
  }
  qc.validationMapConsumables[uuid] = (validationMap[aCmd] || [] ).slice()
  qc.dataMapConsumables[uuid] = (dataMap[aCmd] || [] ).slice()
  if (errorMap[aCmd]){
    qc.errorMapConsumables[uuid] = (errorMap[aCmd]).slice()
  }
  return true
}
qc.handleError = function(aCommandArray, requestData
                    , runParallel){
  qc.handleRequest(aCommandArray, requestData
                    , null, runParallel)
}

/*  
* The dispatchToValCF function is responsible for executing
* any validation functions associated with a specific command.
*/
qc.dispatchToValCF = function(command, parameters){
  return qc.dispatchToCF(command, parameters, qc.validationMapConsumables)
}
/*  
* The dispatchToDCF function is responsible for executing
* any validation functions associated with a specific command.
*/
qc.dispatchToDCF = function(command, parameters){
    //console.log('doing dcfs: '+parameters.stackID)
    return qc.dispatchToCF(command, parameters, qc.dataMapConsumables)
}
/*  
* The dispatchToValCF function is responsible for executing
* any validation functions associated with a specific command.
* The stack is cloned here since it will run in the main thread/page.
*/
qc.dispatchToVCF = function(command, parameters){
  //console.log('validation id:'+parameters.stackID)
  qc.viewMapConsumables[parameters.stackID] = (viewMap[command] || [] ).slice()
  return qc.dispatchToCF(command, parameters, qc.viewMapConsumables)
}
qc.dispatchToCF = function(command, parameters, map){
	var commandFuncs = map[parameters.stackID];
	if(commandFuncs){
		var numFuncs = commandFuncs.length;
		for(var i = 0; i < numFuncs; i++){
      var theControlFunc = commandFuncs.shift()
      theControlFunc.qc_stack_command = command
      theControlFunc.qc_stack_parameters = parameters
			var retVal = theControlFunc(parameters)
			if(retVal == qc.STACK_EXIT){
				return qc.STACK_EXIT
			}
      else if(retVal == qc.STACK_WAIT){
        return qc.STACK_WAIT
      }
		}
	}
	return qc.STACK_CONTINUE;
}

/*
 * These mapping functions require functions for the business rules,
 * view controls, error controls, and security controls NOT the names 
 * of these items as strings.
 */


qc.mapCommandToDCF = function(aCmd, aDCF){
    var funcArray = dataMap[aCmd]
    if(funcArray == null){
        funcArray = new Array()
        dataMap[aCmd] = funcArray
    }
    funcArray.push(aDCF)
}

qc.mapCommandToVCF = function(aCmd, aVCF){
    var funcArray = viewMap[aCmd]
    if(funcArray == null){
        funcArray = new Array()
        viewMap[aCmd] = funcArray
    }
    funcArray.push(aVCF)
}


qc.mapCommandToECF = function(anErrorCmd, anECF){
    var funcArray = errorMap[anErrorCmd]
    if(funcArray == null){
        funcArray = new Array()
        errorMap[anErrorCmd] = funcArray
    }
    funcArray.push(anECF)
}

qc.mapCommandToValCF = function(aCmd, aValCF){
    var funcArray = validationMap[aCmd]
    if(funcArray == null){
        funcArray = new Array()
        validationMap[aCmd] = funcArray
    }
    funcArray.push(aValCF)
}

/*
* this method has a first parameter that is the new command
* and an array of commands that are subcommands to be
* associated under the parent command
*/

qc.mapCommandToSubCommands = function(command, subCommands){
  if(!command || arguments.length < 2){
    return qc.missingCommandError
  }
  var numArgs = arguments.length
  for(var i = 1; i < numArgs; i++){
    var success = qc.addSubCommand(command, arguments[i])
    if(success == qc.missingCommandError){
      groupMap[command] = null
      return qc.missingCommandError
    }
  }
}

qc.addSubCommand = function(aCommand, aSubCommand){
    if(aCommand == null || aSubCommand == null){
      return qc.missingCommandError
    }
    var subCommandSubs = groupMap[aSubCommand]
    if(subCommandSubs){
      var numSubSubs = subCommandSubs.length
      for(var i = 0; i < numSubSubs; i++){
        qc.addSubCommand(aCommand, subCommandSubs[i])
      }
    }
    else{
      var aGroup = groupMap[aCommand]
      if(!aGroup){
        aGroup = new Array()
        groupMap[aCommand] = aGroup
      }
      aGroup.push(aSubCommand)
    }
    return qc.success;
}

//remove white space from the beginning and end of a string
qc.trim = function(aString){
    return aString.replace(/^\s+|\s+$/g, '')
}

//replace all occurances of a string with another
qc.replaceAll = function(aString, replacedString, newSubString){
    while(aString.indexOf(replacedString) > -1){
        aString = aString.replace(replacedString, newSubString)
    }
    return aString
}

//stop an event from doing its' default behavior
qc.stopDefault = function(event){
    if(event){
        event.preventDefault()
    }
    return false
}

/*
 * errorMessage creates a single string message
 * from all of the data available from an error
 * created by a try/catch statement.  If it is 
 * passed something other than an error it 
 * treats it like a string.
 */
qc.errorMessage = function(err){
    var errMsg = 'ERROR: ';
    if(err.name && err.message && err.line && err.sourceURL){
        var fileName = err.sourceURL;
        var stringArr = fileName.split('/');
        var fileName = stringArr[stringArr.length -1];
        errMsg += err.name+': '+err.message+' '+fileName+' line: '+err.line;
    }
    else{
        errMsg += err;
    }
    var curr  = arguments.callee.caller,
    FUNC  = 'function', ANON = "{anonymous}",
    fnRE  = /function\s*([\w\-$]+)?\s*\(/i, stack = [],j=0, fn,args,i
                                        
    errMsg += '\ncall stack:\n';
    while (curr) {
        var fn    = fnRE.test(curr.toString()) ? RegExp.$1 || ANON : ANON;
        var args  = stack.slice.call(curr.arguments);
        var i     = args.length;
        
        while (i--) {
           switch (typeof args[i]) {
                case 'string'  : args[i] = '"'+args[i].replace(/"/g,'\\"')+'"';
                        break;
                case 'function': args[i] = FUNC; 
                        break;
            }
        }
       if(fn != 'logError'){
           errMsg += fn + '(' + args.join() + ')';
       }
        errMsg+="\n";
        curr = curr.caller;
    }
   return errMsg;
}
